hnn_core/network.py

```diff
@@ -389,8 +393,21 @@ def __init__(self, params, add_drives_from_params=False,
         self.pos_dict = dict()
         self.cell_types = dict()
 
-        self._N_pyr_x = self._params['N_pyr_x']
-        self._N_pyr_y = self._params['N_pyr_y']
+        if mesh_shape is None:
+            self._N_pyr_x = self._params['N_pyr_x']
+            self._N_pyr_y = self._params['N_pyr_y']
+        else:
+            _validate_type(mesh_shape, tuple, 'mesh_shape')
+            _validate_type(mesh_shape[0], int, 'mesh_shape[0]')
+            _validate_type(mesh_shape[1], int, 'mesh_shape[1]')
+
+            if mesh_shape[0] < 1 or mesh_shape[1] < 1:
+                raise ValueError('mesh_shape must be a tuple of positive '
+                                 f'integers, got: {mesh_shape}')
+
+            self._N_pyr_x = mesh_shape[0]
+            self._N_pyr_y = mesh_shape[1]
+
```

Comment:
Would it make sense to abstract this into a method like `set_mesh_shape` similar to the `set_cell_positions` method? That way it can be set after initialization to another shape? The method would also call `set_cell_positions` internally so that the positions will be updated if the shape is changed.

Reply:
The main issue is that we don't have a good way to update `net.connectivity` if the shape is changed. It's not impossible but being able to change the shape after initiialization will be a much more involved PR.

Reply:
I agree with @ntolley. It'd be great to have `set_mesh_shape` in the future, but the minimal working solution here is to just set the size/shape of the 2D pyramidal cell grid/mesh.

Reply:
Ok sounds good. Just for my info, at what point does `net.connectivity` get populated?

Reply:
`net.connectivity` gets populated in two places, first in `network_models.py` when the user creates a network instance from one of our templates (e.g., `net = jones_2009_model()`), and second any time an exogenous drive is added to the network (e.g., with `net.add_evoked_drive()`). In both cases, a given connection type is added by calling `net.add_connection()`.